File: android/feature/library/src/main/java/com/mplayeraudio/feature/library/MusicLibraryState.kt
package com.mplayeraudio.feature.library

import com.mplayeraudio.core.domain.musiclibrary.PlaylistId
import com.mplayeraudio.core.domain.musiclibrary.PlaylistRole
import com.mplayeraudio.core.ui.components.MultiplayerCardSurfaceStyle

data class MusicLibraryState(
    val isLoading: Boolean = true,
    val content: MusicLibraryContent? = null,
)

data class MusicLibraryContent(
    val cards: List<MusicLibraryCard>,
)

sealed interface MusicLibraryCard {
    val key: String

    data class Playlist(
        val id: PlaylistId,
        val title: String,
        val trackCount: Int,
        val role: PlaylistRole,
        val layout: PlaylistCardLayout,
        val style: MultiplayerCardSurfaceStyle,
        val artwork: PlaylistCardArtwork,
        val artworkSeed: Int,
    ) : MusicLibraryCard {
        override val key: String = id.toString()
    }
}

enum class PlaylistCardLayout {
    Compact,
    Featured,
}

enum class PlaylistCardArtwork {
    Default,
    Favourites,
}

File: ios/Feature/Library/Sources/LibraryFeature/Presentation/MusicLibraryState.swift
import CoreDomain
import CoreUI
import Foundation

public struct MusicLibraryState: Sendable {
    public var isLoading: Bool
    public var content: MusicLibraryContent?
    public var selectedPlaylist: MusicLibraryDestination?

    public init(
        isLoading: Bool = true,
        content: MusicLibraryContent? = nil,
        selectedPlaylist: MusicLibraryDestination? = nil
    ) {
        self.isLoading = isLoading
        self.content = content
        self.selectedPlaylist = selectedPlaylist
    }
}

public struct MusicLibraryContent: Sendable {
    public let cards: [MusicLibraryCard]

    public init(cards: [MusicLibraryCard]) {
        self.cards = cards
    }
}

public enum MusicLibraryCard: Identifiable, Hashable, Sendable {
    case playlist(PlaylistCard)

    public var id: String {
        switch self {
        case let .playlist(playlist):
            return "\(playlist.ref.provider.renderedKey):\(playlist.ref.id.renderedKey)"
        }
    }

    var span: LibraryGridSpan {
        switch self {
        case let .playlist(playlist):
            playlist.layout == .featured ? .fullWidth : .compact
        }
    }
}

public struct PlaylistCard: Hashable, Sendable {
    public let ref: PlaylistRef
    public let title: String
    public let trackCount: Int
    public let role: PlaylistRole
    public let layout: PlaylistCardLayout
    public let style: MultiplayerCardSurfaceStyle
    public let artwork: PlaylistCardArtwork
    public let artworkSeed: Int

    public init(
        ref: PlaylistRef,
        title: String,
        trackCount: Int,
        role: PlaylistRole,
        layout: PlaylistCardLayout,
        style: MultiplayerCardSurfaceStyle,
        artwork: PlaylistCardArtwork,
        artworkSeed: Int
    ) {
        self.ref = ref
        self.title = title
        self.trackCount = trackCount
        self.role = role
        self.layout = layout
        self.style = style
        self.artwork = artwork
        self.artworkSeed = artworkSeed
    }
}

public enum PlaylistCardLayout: Hashable, Sendable {
    case compact
    case featured
}

public enum PlaylistCardArtwork: Hashable, Sendable {
    case `default`
    case favourites
}

public struct MusicLibraryDestination: Hashable, Sendable, Identifiable {
    public let ref: PlaylistRef
    public let title: String
    public let role: PlaylistRole

    public init(ref: PlaylistRef, title: String, role: PlaylistRole) {
        self.ref = ref
        self.title = title
        self.role = role
    }

    public var id: String {
        "\(ref.provider.renderedKey):\(ref.id.renderedKey)"
    }
}

public enum MusicLibraryAction: Sendable {
    case playlistTapped(PlaylistRef)
    case dismissPlaylistDetail
}

extension PlaylistId {
    var renderedKey: String {
        "\(ownerId.rawValue):\(kind.rawValue)"
    }
}

private extension MusicProviderId {
    var renderedKey: String {
        switch self {
        case .device:
            return "device"

        case .yandexMusic:
            return "yandex"
        }
    }
}
